### DISCUSSION TITLE
Holding and Passing Custom Data between Custom Operators

### DISCUSSION THREAD

**Discussion by karthikeyann** [facebookincubator/velox#11835] [Ideas]
To pass Custom Data (libcudf dataframe  https://docs.rapids.ai/api/libcudf/stable/classcudf_1_1table ) between Custom Operators, I would like design review for a proposed custom vector type `CudfVector`, 
- `CudfVector` class derived from `RowVector` holds the libcudf GPU table as shared_ptr and passed around operators as `RowVectorPtr`
- Custom Cudf operators will `dynamic_pointer_cast<CudfVector>` and use it.
- New operators `toVeloxVector`  and `toCudfVector` will convert data to and from velox.
- Only custom cudf operators are expected to get this custom datatype. Driver adapter or users will make sure to insert these conversion operators in plan fragment so that only custom operators will get this `CudfVector` 

Is creating a new Vector type from `RowVector` an acceptable design?

**Comment by pedroerp** [comment]
Cc: @majetideepak @oerling @Yuhta @mbasmanova @xiaoxmeng

**Comment by pedroerp** [comment]
@karthikeyann thanks for starting this discussion. Like we discussed earlier, another option could be to specialize the Buffer class within a vector to contain a GPU memory location instead of host memory. What are the trade-offs between these two approaches? I suppose cudf::column has a buffer internally?

**Reply by Yuhta** [reply]
Yes inheritance would introduce strong coupling so any change on `RowVector` could break `CudfVector` in the future. Ideally if you can make it work with GPU memory backed buffers it would be the cleanest way.

**Comment by FelixYBW** [comment]
How close the cudf data format is to Arrow? Is it possible to reuse the arrow/velox conversion? In this way to the other Velox CPU operators, it's just a Arrow data source.

**Reply by karthikeyann** [reply]
It's pretty much same as Arrow. All primitive, timestamp, decimal, string and nested types are supported. Some complex types like Union, opaque, binary and variant of list, string are not supported yet.
